src/graph/visitor/PropertyTrackerVisitor.cpp

```diff
@@ -174,27 +171,78 @@ void PropertyTrackerVisitor::visit(AttributeExpression *expr) {
     return;
   }
   auto &propName = constVal.getStr();
-  static const int kUnknownEdgeType = 0;
   switch (lhs->kind()) {
     case Expression::Kind::kInputProperty:
     case Expression::Kind::kVarProperty: {  // $e.name
       auto *varPropExpr = static_cast<PropertyExpression *>(lhs);
       auto &edgeAlias = varPropExpr->prop();
-      propsUsed_.insertEdgeProp(edgeAlias, kUnknownEdgeType, propName);
+      propsUsed_.insertEdgeProp(edgeAlias, unKnowType_, propName);
       break;
     }
     case Expression::Kind::kSubscript: {  // $-.e[0].name
       auto *subscriptExpr = static_cast<SubscriptExpression *>(lhs);
       auto *subLeftExpr = subscriptExpr->left();
-      if (subLeftExpr->kind() == Expression::Kind::kInputProperty) {
-        auto *inputPropExpr = static_cast<InputPropertyExpression *>(subLeftExpr);
-        auto &edgeAlias = inputPropExpr->prop();
-        propsUsed_.insertEdgeProp(edgeAlias, kUnknownEdgeType, propName);
+      auto kind = subLeftExpr->kind();
+      if (kind == Expression::Kind::kInputProperty || kind == Expression::Kind::kVarProperty) {
+        auto *propExpr = static_cast<PropertyExpression *>(subLeftExpr);
+        auto &edgeAlias = propExpr->prop();
+        propsUsed_.insertEdgeProp(edgeAlias, unKnowType_, propName);
+      } else if (kind == Expression::Kind::kListComprehension) {
+        //  match (src_v:player{name:"Manu Ginobili"})-[e*2]-(dst_v) return e[0].start_year
+        auto *listExpr = static_cast<ListComprehensionExpression *>(subLeftExpr);
+        auto *collectExpr = listExpr->collection();
+        if (collectExpr->kind() == Expression::Kind::kInputProperty) {
+          auto *inputPropExpr = static_cast<InputPropertyExpression *>(collectExpr);
+          auto &aliasName = inputPropExpr->prop();
+          propsUsed_.insertEdgeProp(aliasName, unKnowType_, propName);
+        }
       }
       break;
     }
     case Expression::Kind::kFunctionCall: {  // properties(t3).name
-      // TODO(jmq) determine whether it is a vertex or edge
+      auto *funCallExpr = static_cast<FunctionCallExpression *>(lhs);
+      auto funName = funCallExpr->name();
+      std::transform(funName.begin(), funName.end(), funName.begin(), ::tolower);
+      if (funName != "properties") {
+        break;
+      }
+      auto *argExpr = funCallExpr->args()->args().front();
+      auto kind = argExpr->kind();
+      switch (kind) {
+        case Expression::Kind::kVarProperty:
+        case Expression::Kind::kInputProperty: {
+          //  match (v) return properties(v).name
+          auto *inputPropExpr = static_cast<InputPropertyExpression *>(argExpr);
+          auto &aliasName = inputPropExpr->prop();
+          propsUsed_.insertVertexProp(aliasName, unKnowType_, propName);
```

Comment:
How to known `v` is a vertex or edge?

Reply:
edge is only subscriptExpression or listComprehensionExpression,  can be distinguished by the expression type

Reply:
Is it invalid to use `properties(edge).name`?

Reply:
edge is a KEYWORD, parser error